src/kmergraph.cpp

```diff
@@ -105,12 +97,15 @@ KmerNodePtr KmerGraph::add_node(const prg::Path& p)
     KmerNodePtr n(std::make_shared<KmerNode>(nodes.size(), p)); // create the node
     nodes.push_back(n); // add it to nodes
     sorted_nodes.insert(n);
-    assert(k == 0 or p.length() == 0 or p.length() == k);
+
+    bool path_is_valid = k == 0 or p.length() == 0 or p.length() == k;
+    if (!path_is_valid) {
+        FatalError() << "In KmerGraph::add_node(), the node path is not valid (k is " << k
```

Comment:
Yeah, this is a bit subjective. On one end, describing errors in plain english is good for the user to understand it, but on another end, some of the errors are just implementation details, and the description is not useful for the user, as they won't be able to understand them. For many errors, the description is targeted to developers than to the users. Some errors I think are not necessary, but I am not removing any as of now, to keep consistency with the previous versions. But I also think I did not do this error logging it in the proper way, and thus this has been confusing for you to review. So I guess the best to do is to provide a plain english message for the user, but also a decent error logging for developers (will be using [Boost Stacktrace](https://www.boost.org/doc/libs/1_75_0/doc/html/stacktrace.html)) for that. I know that we can always ask the user for the input, and debug the error in debug executions, but many cases just a look at the stack trace might be enough to know already the issue!

Reply:
[Boost Stacktrace](https://www.boost.org/doc/libs/1_75_0/doc/html/stacktrace.html) does not work on our project :( ... Issue is that it was introduced in `Boost 1.65.0`, but we are stuck at `Boost 1.62.0` because of `GATB` (`GATB` introduces a local copy of `Boost 1.62.0`, and it seems that linking any different boost versions give us linker errors... I also searched how to use two different `Boost` versions in a single project, but this can result in runtime errors, so it is not advisable). I ended up using [backward-cpp](https://github.com/bombela/backward-cpp/) for this, and now we provide a plain english error message and a stack trace that will probably help us debug the issue with the user not even sending us the input, and etc. Do you think it is fine to keep it like this @mbhall88 ? An example in a test case that fails with a bug is below

```
[FATAL ERROR]: Negative binomial parameters are invalid (mean is 0.000000, variance is 1.000000)
Aborting...

DEBUG info (for developers, provide this if opening an issue):
Stack trace (most recent call last):
#18   Object "", at 0xffffffffffffffff, in 
#17   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3a45fd, in _start
#16   Object "/usr/lib/x86_64-linux-gnu/libc-2.31.so", at 0x7f9be93070b2, in __libc_start_main
      Source "../csu/libc-start.c", line 308, in __libc_start_main [0x7f9be93070b2]
#15   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc93ac06, in main
      Source "/home/leandro/git/pandora/test/main.cpp", line 10, in main [0x557abc93ac06]
          7:     core->set_logging_enabled(false);
          8: 
          9:     ::testing::InitGoogleTest(&argc, argv);
      >  10:     return RUN_ALL_TESTS();
         11: }
#14   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc93ad8e, in RUN_ALL_TESTS()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Install/include/gtest/gtest.h", line 2473, in RUN_ALL_TESTS [0x557abc93ad8e]
       2470: int RUN_ALL_TESTS() GTEST_MUST_USE_RESULT_;
       2471: 
       2472: inline int RUN_ALL_TESTS() {
      >2473:   return ::testing::UnitTest::GetInstance()->Run();
       2474: }
       2475: 
       2476: GTEST_DISABLE_MSC_WARNINGS_POP_()  //  4251
#13   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcca6d7f, in testing::UnitTest::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 4925, in Run [0x557abcca6d7f]
#12   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccbfc50, in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2469, in HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> [0x557abccbfc50]
#11   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccc63a9, in bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2433, in HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> [0x557abccc63a9]
#10   Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcca853f, in testing::internal::UnitTestImpl::RunAllTests()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 5338, in RunAllTests [0x557abcca853f]
#9    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9c483, in testing::TestSuite::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2816, in Run [0x557abcc9c483]
#8    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9bd48, in testing::TestInfo::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2684, in Run [0x557abcc9bd48]
#7    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abcc9b35d, in testing::Test::Run()
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2508, in Run [0x557abcc9b35d]
#6    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccbec42, in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2469, in HandleExceptionsInMethodIfSupported<testing::Test, void> [0x557abccbec42]
#5    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abccc4f80, in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
      Source "/home/leandro/.hunter/_Base/7d73237/252be92/0470de9/Build/GTest/Source/googletest/src/gtest.cc", line 2433, in HandleSehExceptionsInMethodIfSupported<testing::Test, void> [0x557abccc4f80]
#4    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc7c7316, in EstimateParameters_FitNegativeBinomial_MeanZero_Death_Test::TestBody()
      Source "/home/leandro/git/pandora/test/estimate_parameters_test.cpp", line 68, in TestBody [0x557abc7c7316]
         65: {
         66:     double mean = 0, variance = 1;
         67:     float p, r;
      >  68:     ASSERT_EXCEPTION(fit_negative_binomial(mean, variance, p, r), runtime_error,
         69:         "Negative binomial parameters are invalid");
         70: }
#3    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc4cfdc7, in fit_negative_binomial(double&, double&, float&, float&)
      Source "/home/leandro/git/pandora/src/estimate_parameters.cpp", line 51, in fit_negative_binomial [0x557abc4cfdc7]
         48: {
         49:     bool negative_binomial_parameters_are_ok = mean > 0 and variance > 0 and mean < variance;
         50:     if (!negative_binomial_parameters_are_ok) {
      >  51:         fatal_error("Negative binomial parameters are invalid (mean is ", mean,
         52:             ", variance is ", variance, ")");
         53:     }
         54:     p = mean / variance;
#2    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc4d5428, in void fatal_error<char [51], double, char [15], double, char [2]>(char const (&) [51], double const&, char const (&) [15], double const&, char const (&) [2])
      Source "/home/leandro/git/pandora/include/fatal_error.h", line 36, in fatal_error<char [51], double, char [15], double, char [2]> [0x557abc4d5428]
         33: template<typename... Args>
         34: void fatal_error(Args const&... args) {
         35:     backward::StackTrace stacktrace;
      >  36:     stacktrace.load_here(32);
         37:     backward::Printer stacktrace_printer;
         38:     stacktrace_printer.address = true;
         39:     stacktrace_printer.object = true;
#1    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3a58bd, in backward::StackTraceImpl<backward::system_tag::linux_tag>::load_here(unsigned long, void*, void*)
      Source "/home/leandro/git/pandora/include/backward.hpp", line 859, in load_here [0x557abc3a58bd]
        856:             return 0;
        857:         }
        858:         _stacktrace.resize(depth);
      > 859:         size_t trace_cnt = details::unwind(callback(*this), depth);
        860:         _stacktrace.resize(trace_cnt);
        861:         skip_n_firsts(0);
        862:         return size();
#0    Object "/home/leandro/git/pandora/cmake-build-debug/test/pandora_test", at 0x557abc3aae4c, in unsigned long backward::details::unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback>(backward::StackTraceImpl<backward::system_tag::linux_tag>::callback, unsigned long)
      Source "/home/leandro/git/pandora/include/backward.hpp", line 841, in unwind<backward::StackTraceImpl<backward::system_tag::linux_tag>::callback> [0x557abc3aae4c]
        839: template <typename F> size_t unwind(F f, size_t depth) {
        840:     Unwinder<F> unwinder;
      > 841:     return unwinder(f, depth);
        842: }
        843: 
        844: } // namespace details
```

Reply:
Great workaround! Looks good to me. Do they have a way to control the verbosity of the stacktrace? We could potentially link that to the `-v` flag? I don't mind leaving it as is though...

Reply:
As per https://github.com/bombela/backward-cpp/#printer , we can remove the `Object` line and the address of the source location (although I think this second one is not advisable...)

Reply:
No worries. Leave it as is :+1: